### PR TITLE
fix: mask password value in DefaultPrompter result display

### PR DESCRIPTION
**Pull request by uchiha-bug-hunter**
Prompt result for a `PasswordPrompt`, before this change:

    expected: <*******> but was: <hunter2>

`executeInputPrompt` returns `DefaultInputResult(input, input, prompt)` and the second field is the display value, so `getDisplayResult()` hands back the clear-text password. `promptInternal` splices that into the `? <message> <answer>` header line, which `reader.printAbove` writes before the next prompt and `close()` paints and leaves on the terminal, so a password that was masked while typing is echoed back and stays in scrollback. The older console-ui `AbstractPrompt` keeps a parallel masked `displayBuffer` for exactly this reason.

The fix builds the display value from the prompt's mask inside `executeInputPrompt` (the only site that reads masked input): the mask character repeated for the input length, or empty when `PasswordPrompt.showMask()` is false. Unmasked prompts (`mask == null`) are unchanged, and `getInput()`/`getResult()` still return the real value.

## Summary by CodeRabbit

* **New Features**
  * Password prompts now mask entered characters by default when no mask is specified.
  * Hidden password mode conceals entered characters while preserving the input for processing.
  * Masked results and string representations no longer expose password text.
  * Unmasked prompts continue displaying entered text normally.

* **Documentation**
  * Clarified the difference between typing-time masking and post-input display behavior.

### CODE REVIEW

**Review by gnodet — APPROVED** [review]
Clean, well-targeted security fix that prevents password leakage through the display result. The implementation correctly masks the display value for password prompts while preserving the actual password for callers, with appropriate test coverage.

**Highlights:**
- The `maskedDisplay` method correctly handles all four cases: no mask (regular input), null input, `showMask=false` (empty string), and masked password (mask char repeated). The `instanceof PasswordPrompt` guard appropriately gates the `showMask()` call.
- Good alignment with the older `console-ui` module's `AbstractPrompt`, which maintained a separate `displayBuffer` for the same reason.
- Test coverage is solid: `maskedDisplayNeverLeaksPassword` verifies both the API result and the terminal output stream, and `hiddenMaskShowsNothing` covers the `showMask=false` path.

**Minor notes (not blocking):**
- Pre-existing: `DefaultInputResult.toString()` returns the raw input, which could leak passwords if logged. Worth a follow-up.
- `mask.charValue()` on line 547 could be simplified to just `mask` (auto-unboxing), but this is purely stylistic.

LGTM 👍

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

File: prompt/src/main/java/org/jline/prompt/impl/DefaultPrompter.java [hunk 1]
```diff
@@ -529,6 +529,26 @@ private InputResult executePasswordPrompt(List<AttributedString> header, Passwor
...
+     * the typed text must never be surfaced: the display value is the mask character repeated for
+     * the length of the input, or empty when the prompt hides its input entirely.
+     */
+    private static String maskedDisplay(InputPrompt prompt, String input) {
+        Character mask = prompt.getMask();
+        if (mask == null || input == null) {
+            return input;
+        }
+        if (prompt instanceof PasswordPrompt && !((PasswordPrompt) prompt).showMask()) {
+            return "";
```
[the remark this answers is not in the document]
Comment (uchiha-bug-hunter) [reply prompt/src/main/java/org/jline/prompt/impl/DefaultPrompter.java]: Good point, the interface does document null as "use the default '*'" even though DefaultPasswordPrompt fills it in itself. Fixed in 633bcaf: maskedDisplay now applies the default before deciding the display value, with a test using a custom PasswordPrompt that leaves getMask() null.

File: prompt/src/test/java/org/jline/prompt/PasswordPromptMaskingTest.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+    private static Map<String, ? extends PromptResult<? extends Prompt>> runPassword(
+            String typed, boolean showMask, ByteArrayOutputStream out) throws Exception {
+        PipedInputStream in = new PipedInputStream();
+        PipedOutputStream outIn = new PipedOutputStream(in);
+        outIn.write((typed + "\n").getBytes(StandardCharsets.UTF_8));
+
+        Terminal terminal =
+                TerminalBuilder.builder().type("ansi").streams(in, out).build();
+        terminal.setSize(Size.of(160, 80));
```
[the remark this answers is not in the document]
Comment (uchiha-bug-hunter) [reply prompt/src/test/java/org/jline/prompt/PasswordPromptMaskingTest.java:43]: Added in 633bcaf, both tests now assert the captured output doesn't contain the typed password. The fixture needed the line discipline's own echo turned off first: the piped input was echoed to the output stream before readLine entered raw mode, so the exclusion assertion was tripping on the fixture's echo rather than anything the prompter wrote.

**Comment by uchiha-bug-hunter** [comment]
Pushed 633bcaf addressing the review notes: maskedDisplay now applies the documented '*' default when a PasswordPrompt leaves getMask() null, the charValue() call is gone, and both masking tests assert the captured terminal output never contains the typed password (the fixture disables the line discipline echo first, since the piped input was being echoed before readLine entered raw mode). Prompt module suite is green.

Agreed that DefaultInputResult.toString() returning the raw input is worth fixing, it predates this change so I'll put up a separate PR for it.

**Review by gnodet — APPROVED** [review]
Re-reviewed after new commit `633bcaff`. The follow-up cleanly addresses all previously raised concerns:

- **Null-mask security gap fixed**: `maskedDisplay` now correctly defaults to `'*'` when `PasswordPrompt.getMask()` returns null, consistent with the interface's documented contract.
- **Improved test assertions**: Terminal output stream is now checked to never contain the typed password (with ECHO flag disabled to avoid line discipline contamination).
- **`charValue()` simplification**: Replaced with auto-unboxing, safe since the null check occurs earlier.

The new `nullMaskPasswordPromptStillMasksDisplay` test exercises the exact null-mask code path with a custom anonymous `PasswordPrompt`. SonarCloud passes with zero issues.

LGTM — all edge cases in `maskedDisplay` are correctly handled. ✅

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

File: prompt/src/test/java/org/jline/prompt/PasswordPromptMaskingTest.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+            public Function<String, Boolean> getValidator() {
+                return null;
+            }
+        };
+        Map<String, ? extends PromptResult<? extends Prompt>> results =
+                prompter.prompt(Collections.emptyList(), Collections.<Prompt>singletonList(prompt));
+
+        InputResult result = (InputResult) results.get("pw");
+        assertEquals(secret, result.getInput());
+        assertEquals("*******", result.getDisplayResult());
```
[the remark this answers is not in the document]
Comment (uchiha-bug-hunter) [reply prompt/src/test/java/org/jline/prompt/PasswordPromptMaskingTest.java:164]: Right, the display fix alone left the typing path open: executeInputPrompt handed the null mask straight to readLine, which echoes. Pushed 5f0669c applying the '*' default before readLine, and the null-mask test now disables the line discipline echo before queuing input and asserts the captured output contains the mask and never the password. Kept the TerminalBuilder fixture the other two tests use rather than switching to LineDisciplineTerminal, they're equivalent for this. Prompt suite is green.

**Review by gnodet — APPROVED** [review]
Re-reviewed after new commit `5f0669c7`.

This commit closes a genuine security gap: previously, `maskedDisplay` correctly masked the post-input result, but `readLine` itself was still called with a null mask — meaning the password echoed in cleartext while the user typed. The fix normalizes null to `'*'` *before* `readLine`, consistent with `PasswordPrompt.getMask()` Javadoc.

The test now verifies both levels: terminal output contains `"*******"` and never contains the cleartext password, with ECHO disabled to avoid line discipline contamination.

Clean, minimal, correct. ✅

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

**Review by gnodet — APPROVED** [review]
Clean, well-targeted security fix that prevents password leakage through the display result. The implementation correctly masks the display value at two distinct points (during-typing `readLine` mask and post-input display result).

The `maskedDisplay` helper is well-structured: null-input handling, `showMask=false` short-circuit, null-mask default to `'*'`, and transparent passthrough for non-password prompts. Test coverage exercises masked, hidden, and null-mask scenarios with both API-level assertions and terminal output stream verification.

Minor observations (non-blocking):
- Pre-existing: `DefaultInputResult.toString()` returns `"InputResult{input='" + input + "'}"`, which would expose the raw password if the result object is logged. A follow-up to override `toString()` for password results would close that vector.
- The `showMask=false` path controls only the post-input display value (returns `""`), but `readLine` still receives `'*'` as the mask. Users who want no echo during typing should use `mask('\0')`. Documenting this distinction in `PasswordPrompt.showMask()` Javadoc would help.

### 📋 PR Metadata

| Aspect | Current | Suggested |
|--------|---------|----------|
| Labels | _(none)_ | `bug` |
| Milestone | _(none)_ | `4.4.0` |

### 🔀 Backport Status

⚠️ This security fix targets `master` but no backport PR was found for:
- `4.0.x` — the vulnerable `new DefaultInputResult(input, input, prompt)` pattern exists there

The 3.x branches do not have the prompt module, so no backport is needed there.

---
_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

**Review by gnodet — APPROVED** [review]
Re-reviewed after maintainer's follow-up commit (SonarCloud fixes + review observations squashed onto this branch). All changes are correct and well-scoped:

1. `defaultValue != null ? defaultValue : null` → `defaultValue` — removes no-op ternary (S3776)
2. Comment reword avoids S125 false positive
3. `toString()` now uses `displayInput` for PasswordPrompt — plugs the logging/debug password leak
4. `showMask()` Javadoc clarified for post-input vs typing-time behavior
5. New `toStringNeverLeaksPassword` test validates the toString fix

### 📋 PR Metadata

| Aspect | Current | Suggested |
|--------|---------|----------|
| Labels | _(none)_ | + `bug` |
| Milestone | _(none)_ | `4.4.0` |

### 🔀 Backport Status

⚠️ This security fix targets `master` but no backport PR was found for:
- `4.0.x` — prompt module has same bugs (raw password in displayInput/toString, no null-mask defaulting)

✅ `jline-3.x` — not applicable (prompt module doesn't exist there)

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

**Comment by uchiha-bug-hunter** [comment]
Saw the follow-up commit. Since 7dba073 already fixes DefaultInputResult.toString(), I'll drop the separate PR I had planned for that. Can also cherry-pick this onto 4.0.x once it lands, the same raw displayInput pattern is there.
